mlflow/types/utils.py

```diff
@@ -79,7 +80,69 @@ def clean_tensor_type(dtype: np.dtype):
     return dtype
 
 
-def _infer_datatype(data) -> DataType:
+def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
+    """
+    Infer an MLflow Colspec type from the dataset.
+
+    :param data: data to infer from.
+    :return: Object
+    """
+    dtype = _infer_datatype(data)
+    _validate_inferred_type(dtype)
+    return dtype
+
+
+def _infer_datatype(data: Any) -> Union[DataType, Array, Object]:
+    if isinstance(data, dict):
+        properties = []
+        for k, v in data.items():
+            properties.append(Property(name=k, dtype=_infer_datatype(v)))
+        return Object(properties=properties)
+
+    if isinstance(data, (list, np.ndarray)):
+        return _infer_array_datatype(data)
+
+    return _infer_scalar_datatype(data)
+
+
+def _infer_array_datatype(data: Union[List, np.ndarray]) -> Optional[Array]:
+    """Infer schema from an array. This tries to infer type if there is at least one
+    non-null element in the list, assuming the list has a homogeneous type. However,
+    if the list is empty or all elements are null, returns None as a sign of undetermined.
+
+    :param data: data to infer from.
+    :return: Array(dtype) or None if undetermined
+    """
+
+    dtypes = [_infer_datatype(v) for v in data if not _is_none_or_nan(v)]
+    dtypes = [t for t in dtypes if t is not None]
```

Comment:
/ai Is the following code more efficient/performant than the original one?

```suggestion
    dtypes = (_infer_datatype(v) for v in data if not _is_none_or_nan(v))
    dtypes = [t for t in dtypes if t is not None]
```

Reply:
Good catch thx, indeed I feel like keeping dtypes in a single list itself is not a good idea, probably go back to previous version that uses loop and keep first element alone.